### ISSUE TITLE
[Bug]: Calendar invite emails do not respect time zone

### ISSUE THREAD

**Issue by micah** [nextcloud/server#33171]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

I work with users who live in many different time zones. My timezone is set in the calendar to my personal one, and if I make a meeting and invite someone who has their timezone set to a different timezone, the meeting invite email is sent to them with the time indicated in *my* timezone (without indicating what timezone it is). This is incredibly confusing to people, because they think the time that is in the invite email is in their time zone, and have no indication otherwise.

I did consider if this was a feature request or a bug, but I think that this is actually a bug, but please feel free to re-classify if you disagree.

### Steps to reproduce

1. Set your timezone to X
2. Set up a second user (`bar`) who has a timezone of Y, configure their email
3. Setup a calendar entry and invite `bar` to the meeting
4. Look at the email that is sent inviting you to the meeting

### Expected behavior

The calendar invite email should either convert the timezone to the receiving user's configured timezone in the email or at minimum indicate the TZ in the email. The former would be obviously preferred.

### Installation method

Manual installation

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 7.4

### Web server

Nginx

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "xxx",
            "xxx.xxx"
        ],
        "enable_previews": true,
        "preview_max_x": 1024,
        "preview_max_y": 768,
        "preview_max_scale_factor": 1,
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\Movie",
            "OC\\Preview\\PDF",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\MarkDown"
        ],
        "filesystem_check_changes": 0,
        "activity_expire_days": 14,
        "knowledgebaseenabled": false,
        "auth.bruteforce.protection.enabled": true,
        "blacklisted_files": [
            ".htaccess",
            "Thumbs.db",
            "thumbs.db"
        ],
        "cron_log": true,
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "xxx",
        "htaccess.RewriteBase": "\/nextcloud",
        "dbtype": "mysql",
        "maintenance": false,
        "version": "24.0.2.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "filelocking.enabled": "true",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "timeout": 0
        },
        "log_type": "file",
        "logfile": "\/srv\/nextcloud_data\/nextcloud.log",
        "loglevel": 1,
        "log_rotate_size": 104857600,
        "theme": "",
        "trashbin_retention_obligation": "auto, 7",
        "updater.release.channel": "stable",
        "twofactor_enforced": "false",
        "twofactor_enforced_groups": [],
        "twofactor_enforced_excluded_groups": [],
        "mail_smtpdebug": true,
        "mail_smtpmode": "smtp",
        "mail_smtpsecure": "ssl",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "PLAIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "app_install_overwrite": [
            "twofactor_totp",
            "onlyoffice"
        ],
        "config_is_read_only": "True"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.10.0
  - activity: 2.16.0
  - calendar: 3.4.2
  - circles: 24.0.0
  - cloud_federation_api: 1.7.0
  - comments: 1.14.0
  - contacts: 4.1.1
  - contactsinteraction: 1.5.0
  - dashboard: 7.4.0
  - dav: 1.22.0
  - federatedfilesharing: 1.14.0
  - federation: 1.14.0
  - files: 1.19.0
  - files_pdfviewer: 2.5.0
  - files_rightclick: 1.3.0
  - files_sharing: 1.16.2
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - files_videoplayer: 1.13.0
  - forms: 2.5.1
  - impersonate: 1.11.0
  - logreader: 2.9.0
  - lookup_server_connector: 1.12.0
  - nextcloud_announcements: 1.13.0
  - notifications: 2.12.0
  - oauth2: 1.12.0
  - onlyoffice: 7.5.2
  - password_policy: 1.14.0
  - photos: 1.6.0
  - polls: 3.7.0
  - privacy: 1.8.0
  - provisioning_api: 1.14.0
  - recommendations: 1.3.0
  - serverinfo: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - spreed: 14.0.2
  - support: 1.7.0
  - systemtags: 1.14.0
  - text: 3.5.1
  - theming: 1.15.0
  - twofactor_backupcodes: 1.13.0
  - twofactor_totp: 6.4.0
  - twofactor_webauthn: 0.3.1
  - updatenotification: 1.14.0
  - user_status: 1.4.0
  - viewer: 1.8.0
  - weather_status: 1.4.0
  - workflowengine: 2.6.0
Disabled:
  - admin_audit: 1.13.0
  - bruteforcesettings: 2.4.0
  - encryption
  - files_external
  - firstrunwizard: 2.6.0
  - survey_client: 1.5.0
  - twofactor_u2f: 6.3.0
  - user_ldap
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
Logs disabled
```

**Comment by szaimen** [comment]
Hi, please update to 24.0.9 or better 25.0.3 and report back if it fixes the issue. Thank you!

My goal is to add a label like e.g. 25-feedback to this ticket of an up-to-date major Nextcloud version where the bug could be reproduced. However this is not going to work without your help. So thanks for all your effort!

If you don't manage to reproduce the issue in time and the issue gets closed but you can reproduce the issue afterwards, feel free to create a new bug report with up-to-date information by following this link: https://github.com/nextcloud/server/issues/new?assignees=&labels=bug%2C0.+Needs+triage&template=BUG_REPORT.yml&title=%5BBug%5D%3A+

**Comment by micah** [comment]
On 2023-01-23 06:42:43, Simon L. wrote:
> Hi, please update to 24.0.9 or better 25.0.3 and report back if it fixes the issue. Thank you!

Hello, I'm running 25.0.3 and the issue still exists.


-- 
        micah

**Comment by SebastianKrupinski** [comment]
Fixed as of version 27.

**Comment by micah** [comment]
@SebastianKrupinski are you saying it was fixed in Nextcloud 27?

**Comment by SebastianKrupinski** [comment]
@micah So the emails should state the timezone that the invite was created with in the when line when the event is a partial day event eg. (from 10am to 11am America/Toronto).

As for sending emails with a converted timezone of the recipient, this would only work for users that are created in NC, and would NOT work if you invite a person with an email address. It would also be possible to use the contacts module to check if there is a contact with the same email address and a configured timezone, BUT this would break if there is more then one contact with the same email address configured. 

To give you some insight on how the invitation emails module works internally, ... this module works strictly on email addresses and does not know anything about created users or what contacts exist in the contacts module, so at the moment it can only send invitation emails with the information contained inside the calendar event (Start Date/Time/Time zone and End Date/Time/Time zone and attendee emails)

**Comment by micah** [comment]
@SebastianKrupinski The emails do state the a time and time zone. However, this bug was about the problem that the time sent, and the localization, is in the timezone and locale of the person who created the event. For example, if someone in Brasil creates an event, and invites someone in Germany, the email the person in Germany contains:

```
/Time:/
ter., 16 de dez. de 2024 13:00 - 14:00 (America/Sao_Paulo)
```

Unless the person in Germany knows what 'ter.' means, or what 'dez' means in Brasilian Portuguese, as well as how to convert from America/Soa_Paulo to their timezone, they are completely lost. For the record 'ter' is the abbreviation for "Tuesday" and "dez" is the abbreviation for December.

Imagine you are working with a team of people from around the world, all who have different locales, and different time zones, this makes these emails completely useless.

It is true that Nextcloud now includes the timezone of the originator of the calendar event, but the calendar invite email should convert the timezone to the receiving user's configured timezone in the email, and if that receiving user is not a user in nextcloud, then obviously there would be no way to determine what the right timezone is, so using the originator's timezone is probably the more sensible option (otherwise UTC would be more universal).

You said:

>As for sending emails with a converted timezone of the recipient, this would only work for users that are created in NC, and would NOT work if you invite a person with an email address. It would also be possible to use the contacts module to check if there is a contact with the same email address and a configured timezone, BUT this would break if there is more then one contact with the same email address configured.

I don't quite understand the breaking scenario you describe. Are you saying that there might be two users in nextcloud who have the same email address configured? That is a weird corner case for sure, and it does appear possible to do that in Nextcloud. I suppose the code should check for that, and just fall-back on the default case (use the originator's timezone, or UTC).

**Comment by SebastianKrupinski** [comment]
> I don't quite understand the breaking scenario you describe. Are you saying that there might be two users in nextcloud who have the same email address configured? That is a weird corner case for sure, and it does appear possible to do that in Nextcloud. I suppose the code should check for that, and just fall-back on the default case (use the originator's timezone, or UTC).

Yes, you can have multiple users with the same email address configured. You can also have multiple contacts with the same address.

>Unless the person in Germany knows what 'ter.' means, or what 'dez' means in Brasilian Portuguese, as well as how to convert from America/Soa_Paulo to their timezone, they are completely lost. For the record 'ter' is the abbreviation for "Tuesday" and "dez" is the abbreviation for "December".

I agree that sending out invitations that are localized to your settings is not the best thing, but at the moment there is no good solution to this. This is why our team, that is distributed over dozens of countries, uses English as their default language, so we are all on the same page.

The other problem is, even if we send out the invitation with the recipients localization, the attached calendar event would still be localized with the organizers localization. This would be a issue with all CalDav clients, like Thunderbird, eMClient, iPhone, Android and even the NC UI.

**Comment by micah** [comment]
Thank you for your reply @SebastianKrupinski !

> The other problem is, even if we send out the invitation with the recipients localization, the attached calendar event would still be localized with the organizers localization. This would be a issue with all CalDav clients, like Thunderbird, eMClient, iPhone, Android and even the NC UI.

I would expect that if the email localization issue was fixed, as discussed, then the attached calendar event would also be generated to match that, or do you think there is some reason why that would not be the case?

**Comment by SebastianKrupinski** [comment]
> Thank you for your reply @SebastianKrupinski !
> 
> > The other problem is, even if we send out the invitation with the recipients localization, the attached calendar event would still be localized with the organizers localization. This would be a issue with all CalDav clients, like Thunderbird, eMClient, iPhone, Android and even the NC UI.
> 
> I would expect that if the email localization issue was fixed, as discussed, then the attached calendar event would also be generated to match that, or do you think there is some reason why that would not be the case?

Yes in a perfect world this would be the case but due to the way Calendar events standard works this is really not possible at the moment.  At the moment the standard only specifies how to handle a Start and End timezone for the whole event it self. There is no way to specify a localization for an attendee. What that means is, if you modify the event with a different timezone and localized wording and then send it to someone, that is technically a different event and should be assigned different UUID's. This would break the attendee participation status responses and cause duplicate events. The other way would be to keep the same UUID, for the different events, that way the participation status work, but then what would happen if an attendee makes a change to the event, the wording on the original event would be overridden. I hope this kind of explain things for you.
